drivers/interrupt_controller: stm32: Complete driver factorization

erwango opened the pull request:
Complete code factorization in stm32 exti drivers.

Tested and validated against all STM32 series

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

tagunil commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -117,88 +118,24 @@ void stm32_exti_enable(int line)
-	} else if (line >= 10 && line <= 15) {
-		irqnum = EXTI15_10_IRQn;
-	} else if (line >= 0 && line <= 4) {
+	} else if {
```
Looks really weird!

erwango replied:
Yes, I missed some code in the pushed commit, fixed Txs

tagunil commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -625,6 +466,9 @@ static void __stm32_exti_connect_irqs(struct device *dev)
 		CONFIG_EXTI_STM32_EXTI15_10_IRQ_PRI,
 		__stm32_exti_isr_15_10, DEVICE_GET(exti_stm32),
 		0);
+#elif defined (ONFIG_SOC_SERIES_STM32F2X) || \
```
CONFIG?

erwango replied:
Oups, forgot to amend before pushing... Fixed, thanks!

erwango commented:
@tagunil , txs for review, this should be fixed now

erwango commented:
recheck

erwango committed:
drivers/interrupt_controller: stm32: Complete driver factorization

Complete code factorization in stm32 exti drivers.


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@tagunil , can you review again?

tagunil commented:
@erwango, it may be outside the scope of this cleanup PR, but why we still treat F2/F4/F7 differently by enabling IRQs for lines >=16 on these series and not enabling on others? Which way it should be done actually?
Otherwise LGTM.

erwango commented:
@tagunil, this is right, support could be enhanced, but this was not the aim of this PR. We can log a ticket for that purpose.

avisconti commented:
@erwango 
quickly tested with microphone and sensor samples for ArgonKey and seems good.

avisconti commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -214,45 +151,16 @@ void stm32_exti_enable(int line)
+		default:
+			return;
 		}
 	}
```
case F2/F4/F7 seems in line with previous code.
But I'm not sure about F1/F3. In these case if line is above 15 what happens?

erwango replied:
@avisconti, I reworked a bit the change to be closer to previous code for L4.
For F1/F3, line > 15, code was obviously wrong, also we missed handling for not supported lines for F2/F4/F7 so reporting an error now.

avisconti replied:
> For F1/F3, line > 15, code was obviously wrong, also we missed handling for not supported lines for F2/F4/F7 so reporting an error now.

OK, got it now.

avisconti reviewed (commented):
It is complex...

avisconti commented:
I also re-tested the two commits on ArgonKey (microphone and sensor samples) and on 
disco_l475_iot1 (with lsm6dsl + trigger sample). It works fine.

avisconti commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -1,21 +1,19 @@
- *
- * Chapter 10.2: External interrupt/event controller (EXTI)
+ * Driver is currently implemented to support following EXTI lines
+ * STM32F0/STM32L0/STM32L4: Lines 0 to 15. Lines > 15 not supported
```
This is F1/F3/L4. Correct?

avisconti commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -1,21 +1,19 @@
- * Chapter 10.2: External interrupt/event controller (EXTI)
+ * Driver is currently implemented to support following EXTI lines
+ * STM32F0/STM32L0/STM32L4: Lines 0 to 15. Lines > 15 not supported
+ * STM32F0/STM32L0/STM32L4: Lines 0 to 15. Lines > 15 are not mapped on an IRQ
```
This is F0/L0. Correct?

erwango replied:
Actually this should be 
 * STM32F1/STM32F3: Lines 0 to 15. Lines > 15 not supported
 * STM32F0/STM32L0/STM32L4: Lines 0 to 15. Lines > 15 are not mapped on an IRQ.
Fixed, txs!

erwango committed:
drivers/interrupt_controller: stm32: Complete driver factorization

Complete code factorization in stm32 exti drivers.
Add return value in case line is not implemented.
Except returned error code, refactor has been done iso-feature
compared to previous code. Hence error is reported only when
support was not available on previous series.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

avisconti reviewed (approved):
OK, now seems fine. Thx for the comment at the top of the file. It helps.
I have also re-tested once more on ArgonKey and it works fine.
